pkg/controller/factory/base_controller.go

```diff
@@ -137,7 +137,10 @@ func (c *baseController) Run(ctx context.Context, workers int) {
 }
 
 func (c *baseController) Sync(ctx context.Context, syncCtx SyncContext) error {
-	return c.sync(ctx, syncCtx)
+	var err error
```

Comment:
This needs to go where you launch the go func

```
func (c *baseController) runWorker(queueCtx context.Context) {
	var workerWaitGroup sync.WaitGroup
	workerWaitGroup.Add(1)
	go func() {
		defer workerWaitGroup.Done()
		for {
			select {
			case <-queueCtx.Done():
				return
			default:
				c.processNextWorkItem(queueCtx)
			}
		}
	}()
	workerWaitGroup.Wait()
}
```

You want a defer utilruntime.HandleCrash() like we do `wait.BackoffUntil`